src/Interpreters/InterpreterInsertQuery.cpp

```diff
@@ -703,6 +709,100 @@ QueryPipeline InterpreterInsertQuery::buildInsertPipeline(ASTInsertQuery & query
 }
 
 
+std::optional<QueryPipeline>
+InterpreterInsertQuery::distributedWriteIntoReplicatedMergeTreeFromClusterStorage(const ASTInsertQuery & query, ContextPtr local_context)
+{
+    if (query.table_id.empty())
+        return {};
+
+    StoragePtr dst_storage = DatabaseCatalog::instance().getTable(query.table_id, local_context);
+    if (!dst_storage->isMergeTree() || !dst_storage->supportsReplication())
+        return {};
+
+    auto & select = query.select->as<ASTSelectWithUnionQuery &>();
+    StoragePtr src_storage;
+    if (select.list_of_selects->children.size() == 1)
+    {
+        if (auto * select_query = select.list_of_selects->children.at(0)->as<ASTSelectQuery>())
+        {
+            JoinedTables joined_tables(Context::createCopy(local_context), *select_query);
```

Comment:
`JoinedTables` is expecting already substituted `with` statements
Need we need smth like 
```
if (local_context->getSettingsRef()[Setting::enable_global_with_statement])
    ApplyWithAliasVisitor::visit(select.list_of_selects->children.at(0));
ApplyWithSubqueryVisitor(local_context).visit(select.list_of_selects->children.at(0));
```